server/chat-plugins/hosts.ts

```diff
@@ -0,0 +1,352 @@
+		html += `<tr><th>Lowest IP address</th><th>Highest IP address</th><th>Name</th><th>Host</th></tr>`;
+		const sortedDatacenters = IPTools.datacenters;
+		sortedDatacenters.sort((a, b) => a.minIP - a.minIP);
+		for (const datacenter of sortedDatacenters) {
+			html += `<tr>`;
+			html += `<td>${IPTools.numberToIP(datacenter.minIP)}</td>`;
+			html += `<td>${IPTools.numberToIP(datacenter.maxIP)}</td>`;
+			html += Utils.html`<td>${datacenter.name}</td>`;
+			html += Utils.html`<td>${datacenter.host}</td>`;
+			html += `</tr>`;
+		}
+		html += `</table></div>`;
+		return html;
+	},
+
+	hosts(query) {
+		function ipSort(a: string, b: string) {
+			let i = 0;
+			let diff = 0;
+			const aParts = a.split('.');
+			const bParts = b.split('.');
+			while (diff === 0) {
+				diff = (parseInt(aParts[i]) || 0) - (parseInt(bParts[i]) || 0);
+				i++;
+			}
+			return diff;
+		}
+
+		this.title = "Hosts";
+		if (!this.can('globalban')) return 'Permission denied.';
+		const type = toID(query[0]) || 'all';
+
+		const openProxies = ['all', 'proxyips', 'proxies'].includes(type) ? [...IPTools.singleIPOpenProxies] : [];
+		const proxyHosts = ['all', 'proxyhosts', 'proxies'].includes(type) ? [...IPTools.proxyHosts] : [];
+		const mobileHosts = ['all', 'mobile'].includes(type) ? [...IPTools.mobileHosts] : [];
+		const residentialHosts = ['all', 'residential', 'res'].includes(type) ? [...IPTools.residentialHosts] : [];
+
+		openProxies.sort(ipSort);
+		proxyHosts.sort();
+		mobileHosts.sort();
+		residentialHosts.sort();
+
+		let html = `<div class="ladder pad"><h2>Hosts:</h2><table>`;
+		html += `<tr><th>Host</th><th>Type</th></tr>`;
+
+		for (const proxyIP of openProxies) {
+			html += `<tr><td>${proxyIP}</td><td>Single IP open proxy</td></tr>`;
+		}
+		for (const proxyHost of proxyHosts) {
+			html += `<tr><td>${proxyHost}</td><td>Proxy host</td></tr>`;
+		}
+		for (const mobileHost of mobileHosts) {
+			html += `<tr><td>${mobileHost}</td><td>Mobile host</td></tr>`;
+		}
+		for (const resHost of residentialHosts) {
+			html += `<tr><td>${resHost}</td><td>Residential host</td></tr>`;
+		}
+		html += `</table></div>`;
+		return html;
+	},
+};
+
+export const commands: ChatCommands = {
+	dc: 'datacenters',
+	datacenter: 'datacenters',
+	datacenters: {
+		'': 'help',
+		help() {
+			return this.parse('/help datacenters');
+		},
+
+		show: 'view',
+		view(target, room, user) {
+			if (!this.can('globalban')) return;
+			return this.parse('/join view-datacenters');
+		},
+		viewhelp: [
+			`/datacenters view - View the list of datacenters. Requires: @ &`,
+		],
+
+		// Originally by Zarel
+		widen: 'add',
+		async add(target, room, user, connection, cmd) {
+			if (!this.can('lockdown')) return false;
+			if (!target) return this.parse('/help datacenters add');
+			// should be in the format: IP, IP, name, URL
+			const widen = (cmd === 'widendatacenters');
+
+			const datacentersToAdd: Datacenter[] = [];
+			for (const row of String(target).split("\n")) {
```

Comment:
I believe our guidelines on stringles would have you use `'\n'` here?

Reply:
Why the string cast? I also, I would use `` `\n` ``.

Reply:
(I'm not sure - this code comes from `admin.ts` and was ostensibly Zarel's work, so I figured there was a reason)